lib/puppet/provider/foreman_smartproxy/rest_v3.rb

```diff
@@ -4,7 +4,7 @@
   def proxy
     @proxy ||= begin
       r = request(:get, 'api/v2/smart_proxies', :search => %{name="#{resource[:name]}"})
-      raise Puppet::Error.new("Proxy #{resource[:name]} cannot be retrieved: #{error_message(r)}") unless success?(r)
```

Comment:
I think this is a degradation in usefulness. Now you no longer know if it was GET or an update action. See `create` and `destroy`.

Reply:
I want to make some kind of change here because the current wording `Proxy #{resource[:name]} cannot be retrieved:` is often misunderstood by users to mean there is an issue with their Foreman proxy infrastructure. I'll try to do this in a way that remains sensitive to the HTTP verb and conveys the same information in a way that is less open to misinterpretation.

Reply:
Perhaps `Proxy #{resource[:name]} cannot be retrieved from Foreman server` is sufficient.

Reply:
The important part is that it's retrieving data (as opposed to creating, updating or removing).

Perhaps it should be qualified as `Smart Proxy` instead of `Proxy`. It would introduce branding problems, but I think that's acceptable. `Proxy` is ambiguous since we also have HTTP proxies and AFAIK we don't brand the API endpoints either.

Reply:
Does the response in this case contain the full problem? e.g. to know if it was an auth error, a cannot find that proxy error, other

Reply:
Looks like the full error is only returned here:
https://github.com/theforeman/foreman/blob/b6c40daf3a4b589630e06d409697a62b664c27f0/app/views/api/v2/errors/unprocessable_entity.json.rabl#L6
So only on HTTP 422.

However, that discussion really belongs in https://github.com/theforeman/puppet-foreman/pull/912/files#r553307082. I insist that it still logs the action it was trying to perform.

Reply:
I've updated such that it logs more clearly what is the exact action it tried to perform. The messages now include the http verb,  uri, a simple explanation of the issue, and the ability to log an additional custom message per http error code.